### PR TITLE
Migrate remaining features into CodeEdit

### LINKED ISSUE #41361: Minor usability issue with triple quotes
**Godot version:**

v3.2.2.stable.official

**OS/device including version:**

MacBookAir8,1
macOS Catalina 10.15.5 (19F101)

**Issue description:**

When I type three quotes in a row for a multiline comment or string, the script editor shouldn't automatically close the last one.

**Steps to reproduce:**
Type three quotes in any script

**Minimal reproduction project:**

### PR DESCRIPTION
**Pull request by Paulb23**
Continuation of #31739

---

Apologies for the larger PR was going to limit this to brace completion, but the remaining features seemed small enough to bundle in here.

After this PR, I consider `CodeEdit` good enough for 4.0 and beyond. Next steps will be cleaning up `TextEdit` and prepping the API similarly, before bug fixes / optimisation. After this point we can consider #31739 done and can create smaller tasks for the remaining parts, as they will not have an affect on the outward facing API.

The only remaining "code like" feature in `TextEdit` is highlighting all occurrences of the selected text. I did not consider this a strong enough feature to consider migrating to `CodeEdit`, though am happy to do so if others think this should be the case.

---

Firstly, this PR migrates auto brace completion. This has been made to use strings rather then single chars to allow handling of larger symbols such as gdscript multiline comments / strings. And has exposed to the API. A new overridable method `handle_unicode_input` has been added, which is called when the user types a char. This way the child class does not need to check all shortcuts before handling the input.

For Brace matching, only the API has been migrated and exposed, until drawing can be somewhat controlled in `CodeEdit`. As such could be classed as "incomplete", in the sense that only hardcoded chars are highlighted and not everything added in auto brace completion.

Secondly, line length guidelines have been moved and exposed. Rather then a soft and hard limit it now takes an array of columns. The first entry is a "hard" limit, the following entries are a "soft" limit and drawn at 50% opacity.

The last feature is symbol lookup, this has been reconfigured and exposed to the API. At the moment, underline drawing is still handled by `TextEdit`, however everything else is in `CodeEdit`

Finally, cleaned up the inspector a little by moving gutters into a sub category and changing a few default values. On top of this finished up all the remaining `CodeEdit` docs.

---

closes #41361

### CODE REVIEW

File: doc/classes/TextEdit.xml [hunk 1]
```diff
@@ -982,7 +984,7 @@
 			Sets the [StyleBox] of this [TextEdit].
 		</theme_item>
 		<theme_item name="outline_size" type="int" default="0">
-			The size of the text outline.
+			The tint of text outline of the [CodeEdit].
```
Comment (KoBeWi) [inline doc/classes/TextEdit.xml:987]: This description is wrong. Tint refers to color, not size.

File: scene/gui/code_edit.cpp [hunk 2]
```diff
@@ -1918,11 +2203,78 @@ void CodeEdit::_bind_methods() {
...
+			if (line[(p_col - 1) - j] != open_key[(open_key.length() - 1) - j]) {
+				is_match = false;
+				break;
+			}
+		}
+
+		if (is_match) {
+			idx = i;
+			break;
+		}
```
Comment (KoBeWi) [inline scene/gui/code_edit.cpp:2246]: This looks like the `idx` variable is not necessary. It's modified only once, so you could just do `return i;` and `return -1;` at the end. Same in the following method.

File: scene/gui/code_edit.h [hunk 3]
```diff
@@ -210,14 +223,26 @@ class CodeEdit : public TextEdit {
...
 
+	/* Line length guidelines */
+	TypedArray<int> line_length_guideline_columns;
+	Color line_length_guideline_color;
+
+	/* Symbol lookup */
+	bool symbol_lookup_on_click_enabled = false;
+
+	String symbol_lookup_new_word = "";
+	String symbol_lookup_word = "";
```
Comment (KoBeWi) [inline scene/gui/code_edit.h:234]: You don't need to initialize Strings.

**Comment by KoBeWi** [comment]
I looked over the code and it's mostly ok. Except maybe comments, we usually use `// Double slash comments.` instead of `/*This thing.*/`, especially for single-line comments. The old comments using `/**/` should be updated too probably.

Also would be nice to have the PR rebased, so it can be tested more easily.

**Comment by KoBeWi** [comment]
Hover underline doesn't seem to work:
phMXjR4kVn

**Comment by Paulb23** [comment]
Yeah I saw that too. It's not caused by this PR as it's the same in `master`.

**Review by KoBeWi — APPROVED** [review]
Well, everything else seems fine (at least in the editor). Code is ok too.

**Comment by akien-mga** [comment]
Thanks!

File: scene/gui/code_edit.h (hunk 3 above)
Comment (akien-mga) [reply scene/gui/code_edit.h:234]: I missed this, but that can be done as a follow-up (maybe a regex search as there might be more unneeded String inits in the codebase).

Comment (akien-mga) [reply scene/gui/code_edit.h:234]: Yeah there's a ton of them:
```
$ rg 'String .* = "";' | wc -l
106
```

**Pull request merged**